sympy/solvers/inequalities.py

```diff
@@ -788,37 +802,43 @@ def _solve_inequality(ie, s, linear=False):
         else:
             p = Poly(expr)
 
-    e = expanded = p.as_expr()  # this is in exanded form
     if rv is None:
-        # Do a safe inversion of e, moving non-s terms
-        # to the rhs and dividing by a nonzero factor if
-        # the relational is Eq/Ne; for other relationals
-        # the sign must also be positive or negative
-        rhs = 0
-        b, ax = e.as_independent(s, as_Add=True)
-        e -= b
-        rhs -= b
-        ef = factor_terms(e)
-        a, e = ef.as_independent(s, as_Add=False)
-        if (a.is_zero != False or  # don't divide by potential 0
-                a.is_negative ==
-                a.is_positive == None and  # if sign is not known then
-                ie.rel_op not in ('!=', '==')): # reject if not Eq/Ne
-            e = ef
-            a = S.One
-        rhs /= a
-        if a.is_positive:
-            rv = ie.func(e, rhs)
-        else:
-            rv = ie.reversed.func(e, rhs)
+        e = p.as_expr()  # this is in expanded form
+        if rv is None:
+            # Do a safe inversion of e, moving non-s terms
+            # to the rhs and dividing by a nonzero factor if
+            # the relational is Eq/Ne; for other relationals
+            # the sign must also be positive or negative
+            rhs = 0
+            b, ax = e.as_independent(s, as_Add=True)
+            e -= b
+            rhs -= b
+            ef = factor_terms(e)
+            a, e = ef.as_independent(s, as_Add=False)
+            if (a.is_zero != False or  # don't divide by potential 0
+                    a.is_negative ==
+                    a.is_positive == None and  # if sign is not known then
+                    ie.rel_op not in ('!=', '==')): # reject if not Eq/Ne
+                e = ef
+                a = S.One
+            rhs /= a
+            if a.is_positive:
+                rv = ie.func(e, rhs)
+            else:
+                rv = ie.reversed.func(e, rhs)
+
     # return conditions under which the value is
     # valid, too.
-    conds = [rv]
+    conds = []
     beginning_denoms = denoms(ie.lhs) | denoms(ie.rhs)
-    current_denoms = denoms(expanded)
+    current_denoms = denoms(rv)
     for d in beginning_denoms - current_denoms:
         conds.append(_solve_inequality(Ne(d, 0), s, linear=linear))
-    return And(*conds)
+    if conds:
```

Comment:
This is the logic that needs to be checked carefully. Also make sure that a test for this nonlinear case is added (i.e. conditions are returned when a nonlinear expression is solved).

Reply:
```
if conds:
    if rv is not S.false:
        conds.append(rv)
    rv = And(*conds)
```
```
if isinstance(expr, Ne) and rv is S.false:
    return Piecewise((rv, And(*conds)), (S.true, True))
elif isinstance(expr, Eq):
    # in case of Eq, if rv is false and conditions are not met the result is still false
    return And(*conds)
else:
    # in other cases, if conditions are not met then a type error should be raised
    return Piecewise((rv, And(*conds)),(Or([*(1/c.lhs < 0) for c in conds]), True))
```

Reply:
On L558 in inequalities.py,  `domain = continuous_domain(e, gen, domain)` should be replaced by `domain = continuous_domain(expanded_e, gen, domain)`. This should resolve some of the test failures.